core/src/main/java/org/infinispan/interceptors/InterceptorStage.java

```diff
@@ -0,0 +1,54 @@
+package org.infinispan.interceptors;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.commands.VisitableCommand;
+import org.infinispan.context.InvocationContext;
+
+/**
+ * Actions that an asynchronous interceptor can perform.
+ *
+ * @author Dan Berindei
+ * @since 9.0
+ */
+public interface InterceptorStage {
+
+   /**
+    * Invoke the next interceptor, possibly with a new command.
+    * <p>
+    * {@link InvocationStage} then allows the caller to add a callback that will be executed after the
+    * remaining interceptors.
+    */
+   InvocationStage invokeNext(InvocationContext ctx, VisitableCommand command);
+
+   /**
+    * Return a value directly, skipping the remaining interceptors in the chain.
+    */
+   BasicInvocationStage returnWith(Object returnValue);
+
+   /**
+    * Suspend the invocation until {@code stageFuture} completes, then evaluate its result and skip the remaining
+    * interceptors.
+    * <p>
+    * The caller is supposed to call the
+    */
+   InvocationStage goAsync(InvocationContext ctx, VisitableCommand command,
+         CompletableFuture<BasicInvocationStage> stageFuture);
+
+   /**
+    * Suspend the invocation until {@code delay} completes, then if successful invoke the next interceptor.
+    * <p>
+    * If {@code delay} completes exceptionally, skip the next interceptor and continue with the exception.
+    */
+   InvocationStage invokeNextAsync(InvocationContext ctx, VisitableCommand command, CompletableFuture<?> delay);
+
+   /**
+    * Suspend the invocation until {@code returnValueFuture} completes, then return its result and skip the remaining
+    * interceptors.
+    * <p>
+    * The caller can continue invoking the next interceptor, e.g.
+    * {@code goAsync2(ctx, command, v).thenApply((rCtx, rCommand, rv, t) -> invokeNext(rCtx, rCommand))}
+    */
+   BasicInvocationStage returnWithAsync(CompletableFuture<?> valueFuture);
```

Comment:
I haven't really thought about `CompletableFuture<Void>`, although I would probably use an unchecked cast instead of `thenApply(nil -> null)`.

I was thinking of when you have a more specific type, like `CompletableFuture<RemoteCacheEntry>`, but looking at the existing interceptor I was able to change the parameter to `CompletableFuture<Object>` without any compilation error.